File: Traditional-Algorithms/scala/leetcode1.scala
// 方法一： 考察to与until的区别，to是[a, b]区间，until是[a, b)区间
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        for(i <- 0 to nums.length - 2; j <- i + 1 to nums.length-1){
            if(nums(i) + nums(j) == target) return Array(i, j) 
        }
        return Array()
    }
}

object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        for(i <- 0 until nums.length - 1; j <- i + 1 until nums.length){
            if(nums(i) + nums(j) == target) return Array(i, j) 
        }
        return Array()
    }
}

File: Traditional-Algorithms/golang/LeetCode1.go
func twoSum(nums []int, target int) []int {
    var s = len(nums)
    var m = map[int]int{}
    for i := 0; i < s; i++ {
        m[nums[i]] = i;
    }

    for i := 0; i < s; i++{
       
        if e, ok := m[target - nums[i]]; ok && i != m[target - nums[i]] {
            return []int{i, e}
        }  
    }
    return nil
}


// 精简版实现
func twoSum(nums []int, target int) []int {
    m := map[int]int{}
    for i := range nums {
        if e, ok := m[target - nums[i]]; ok && i != m[target - nums[i]] {return []int{i, e}}
        m[nums[i]] = i;
    }
    return nil
}


// 初始化map的方式不同
func twoSum(nums []int, target int) []int {
    m := make(map[int]int)  // 和上一种方法初始化map的方式不同
    for i := range nums {
        if e, ok := m[target - nums[i]]; ok && i != m[target - nums[i]] {return []int{i, e}}
        m[nums[i]] = i;
    }
    return nil
}